src/Build/Utilities/FileSpecMatchTester.cs

```diff
@@ -41,31 +48,43 @@ public static FileSpecMatcherTester Parse(string currentDirectory, string fileSp
             return new FileSpecMatcherTester(currentDirectory, unescapedFileSpec, filenamePattern, regex);
         }
 
+        /// <summary>
+        /// Returns true if the given file matches this file spec.
+        /// </summary>
         public bool IsMatch(string fileToMatch)
         {
             Debug.Assert(!string.IsNullOrEmpty(fileToMatch));
 
+            string normalizedFileToMatch = FileUtilities.GetFullPathNoThrow(Path.Combine(_currentDirectory, fileToMatch));
+            return IsMatchNormalized(normalizedFileToMatch);
+        }
+
+        /// <summary>
+        /// Same as <see cref="IsMatch" /> but the argument is expected to be a normalized path.
+        /// </summary>
+        public bool IsMatchNormalized(string normalizedFileToMatch)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(normalizedFileToMatch));
+
             // We do the matching using one of three code paths, depending on the value of _filenamePattern and _regex.
             if (_regex != null)
             {
-                string normalizedFileToMatch = FileUtilities.GetFullPathNoThrow(Path.Combine(_currentDirectory, fileToMatch));
                 return _regex.IsMatch(normalizedFileToMatch);
             }
 
             if (_filenamePattern != null)
             {
                 // Check file name first as it's more likely to not match.
-                string filename = Path.GetFileName(fileToMatch);
+                string filename = Path.GetFileName(normalizedFileToMatch);
                 if (!FileMatcher.IsMatch(filename, _filenamePattern))
                 {
                     return false;
                 }
 
-                var normalizedFileToMatch = FileUtilities.GetFullPathNoThrow(Path.Combine(_currentDirectory, fileToMatch));
                 return normalizedFileToMatch.StartsWith(_currentDirectory, StringComparison.OrdinalIgnoreCase);
```

Comment:
I'm kinda wondering if it's worth optimizing this to use EndsWith instead of StartsWith, as I imagine one of the most common differences is amd64 vs. no amd64. Probably not worth it.

Reply:
I don't believe that `EndsWith` walks the string differently than `StartsWith`. They both use the same underlying comparison routines, just the (sub)strings being compared are different. We need the file path to start with the directory so `StartsWith` should be the right choice. My gut is that a manual implementation that walks the string backwards would be slower on average than the highly optimized BCL code and it is not worth it.

https://github.com/dotnet/runtime/blob/b3ab2eb4db830f6c520340f80651b806fd9dfce8/src/libraries/System.Private.CoreLib/src/System/Globalization/Ordinal.cs#L14

Reply:
Good point 👍